### PR TITLE
[`GPTNeoX`] Flex Attention + Refactor

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Adds flex attention and the refactor according to #34809

However, I discovered several issues in the current version of gemma2 (#34282):
- It seems like that flex attention needs a transpose afterwards like sdpa
- Loading flex attn with from pretrained didn't work and hence, current tests use another attn implementation (eager or sdpa not sure again)
- Tests could gain from similar tests like sdpa :D for now it's a bit of a hassle to always have some integration test added when it could be a more general test for all subsequent models
- I'm not familiar with better transformers or limitations of flex attn --> added some todos in case we need to check in
- Flex attn doesn't support dropout (or maybe I've overlooked something)
- Setting `model.config._attn_implementation = ...` should be tracked somewhere and checked for sanity as done the first time - for now it silently overwrites and could cause some ugly errors (tested with changing to flash attention 2 while not having fa2 installed)
- Documentation should be added somewhere (prolly perf or something else)

So tbh, I'm not sure whether to split this PR into several ones, e.g. a gemma fix, general loading, general tests, docs, and then subsequent models, or not

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Review by vasqu — COMMENTED** [review]
A collection of comments which partially show the issues I listed above

**Comment by vasqu** [comment]
CI failings seem unrelated, flaky tests (e.g. XLM, Qwen2VL)

**Comment by vasqu** [comment]
Possible TODO -> fallback to eager when using head mask in fa2, sdpa + add head mask in flex attention (should be possible via score mod)

**Comment by dame-cell** [comment]
Yes I was testing it for Gemma as well it needed a transpose at the end as well 

If you don't mind could you check the pull request i did for gemma seems I keep failing some tests 

Also the gemma 2 now supports new stuff in the  configuration which confused me a lot

 the attn logit soft capping  
 Also the model.config._attn_implementation is not really implemented correctly for example it does not reference correctly uses the correct attn upon choosing one

**Comment by vasqu** [comment]
@dame-cell I'll take a look tomorrow! I'm busted for today :) 

But as quick thing to let the loading be handled correctly look into my changes into the utils folder and modeling_utils. With those changes, loading should be handled correctly. Tbh, that's one of the main reasons why I think it might be better to split some PRs and get loading etc correctly first before we start adding.

**Comment by dame-cell** [comment]
> @dame-cell I'll take a look tomorrow! I'm busted for today :) 
> 
> But as quick thing to let the loading be handled correctly look into my changes into the utils folder and modeling_utils. With those changes, loading should be handled correctly. Tbh, that's one of the main reasons why I think it might be better to split some PRs and get loading etc correctly first before we start adding.

Hmmm ohh I get it I see thanks for letting me know 😀

**Comment by ArthurZucker** [comment]
Feel free to ping me once you feel like this is ready! 🤗

**Comment by vasqu** [comment]
I think it should be ready @ArthurZucker just found something on the fly just a min ago, should be good to go :smile:

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -123,6 +311,13 @@ def __init__(self, config, layer_idx=None):
         self.is_causal = True
         self.layer_idx = layer_idx
 
+        # Attention specific information for the implementations
+        # TODO: Should be removed once Flash Attention for RoCm is bumped to 2.1.
+        # flash_attn<2.1 generates top-left aligned causal mask, while what is needed here is bottom-right alignement, that was made default for flash_attn>=2.1. This attribute is used to handle this difference. Reference: https://github.com/Dao-AILab/flash-attention/releases/tag/v2.1.0.
+        # Beware that with flash_attn<2.1, using q_seqlen != k_seqlen (except for the case q_seqlen == 1) produces a wrong mask (top-left).
+        self._flash_attn_uses_top_left_mask = not is_flash_attn_greater_or_equal_2_10()
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:325]: Potentially yes, but also because you need a higher version of torch for gemma2 AFAIR!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 2]
```diff
@@ -147,20 +349,82 @@ def forward(
...
+            if output_attentions:
+                warning_msg += " `output_attentions=True`"
+            if output_attentions and head_mask is not None:
+                warning_msg += " and `head_mask` not None"
+            elif head_mask is not None:
+                warning_msg += " `head_mask` not None"
+            warning_msg += f" not supported in {attention_type}"
+
+            logger.warning_once(warning_msg)
+            attention_type = "eager"
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:394]: That's supper verbose 😢 we could just say that we are falling back to eager due to config missmatch or something!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 3]
```diff
@@ -147,20 +342,71 @@ def forward(
...
+
+        if (
+            self.training
+            and self.config.attention_dropout > 0
+            and self.config._attn_implementation == "flex_attention"
+        ):
+            logger.warning_once(
+                f"Setting `attention_type` to `eager` because `dropout` is not supported in {attention_type}"
+            )
+            attention_type = "eager"
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:404]: that's a great catch! but we can add it to the score-mod no?

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 4]
```diff
@@ -123,6 +318,13 @@ def __init__(self, config, layer_idx=None):
         self.is_causal = True
         self.layer_idx = layer_idx
 
+        # Attention specific information for the implementations
+        # TODO: Should be removed once Flash Attention for RoCm is bumped to 2.1.
+        # flash_attn<2.1 generates top-left aligned causal mask, while what is needed here is bottom-right alignement, that was made default for flash_attn>=2.1. This attribute is used to handle this difference. Reference: https://github.com/Dao-AILab/flash-attention/releases/tag/v2.1.0.
+        # Beware that with flash_attn<2.1, using q_seqlen != k_seqlen (except for the case q_seqlen == 1) produces a wrong mask (top-left).
+        self._flash_attn_uses_top_left_mask = not is_flash_attn_greater_or_equal_2_10()
+        self.require_contiguous_qkv = version.parse(get_torch_version()) < version.parse("2.2.0")
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:326]: no this should only be checked in SDPA path!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:325]: And it should only be check in FA2 path!

**Review by ArthurZucker — COMMENTED** [review]
Thanks for improving the API `_check_and_enable_flex_attn` was missing from my inital PR!

**Comment by vasqu** [comment]
I'll take a look tomorrow or the day after :)

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -1799,6 +1809,40 @@ def _check_and_enable_sdpa(cls, config, hard_check_only: bool = False) -> Pretra
...
+        """
+        if hard_check_only:
+            if not cls._supports_flex_attn:
+                # TODO: add contribution notice?
+                raise ValueError(
+                    f"{cls.__name__} does not support an attention implementation through torch's flex_attention."
+                    " If you believe this error is a bug, please open an issue in Transformers GitHub repository"
+                    ' and load your model with the argument `attn_implementation="eager"` meanwhile.'
+                    ' Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="eager")`'
+                )
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/modeling_utils.py]: Added link to the issue

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Added this within FA2 only and made the check an lru cache so we don't suffer from performance downgrades.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Sadly not, in the case of the head mask the order of ops doesn't matter since we completely turn the head to all zeros but dropout still depends on the correct distribution calculations and just then turns off some values.

The order of ops is: `dropout(softmax(score_mod(Q, K)))` --> we would introduce unwanted behaviour.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 6]
```diff
@@ -92,6 +97,189 @@ def _init_weights(self, module):
...
+            f"The input hidden states seems to be silently casted in float32, this might be related to"
+            f" the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+            f" {target_dtype}."
+        )
+
+        query = query.to(target_dtype)
+        key = key.to(target_dtype)
+        value = value.to(target_dtype)
+
+    attention_dropout = attention_dropout if training else 0.0
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: I've added this to the FA2 interface, lmk if this is correct and/or if I should change things.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 7]
```diff
@@ -92,6 +97,196 @@ def _init_weights(self, module):
...
+
+    # Mask heads if we want to
+    if head_mask is not None:
+        attn_weights = attn_weights * head_mask
+
+    attn_weights = nn.functional.dropout(attn_weights, p=attention_dropout, training=training)
+    attn_output = torch.matmul(attn_weights, value)
+
+    # Reshape outputs
+    attn_output = GPTNeoXAttention._merge_heads(attn_output, num_attention_heads, attn_head_size)
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Noticed that we can handle like the other implementations and only introduce a transpose here :D

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 8]
```diff
@@ -92,6 +97,196 @@ def _init_weights(self, module):
...
+    **_kwargs,
+):
+    query_length = query.shape[-2]
+
+    # GPT-neo-X casts query and key in fp32 to apply rotary embedding in full precision
+    value_dtype = value.dtype
+    if query.dtype != value_dtype:
+        query = query.to(value_dtype)
+    if key.dtype != value_dtype:
+        key = key.to(value_dtype)
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Done

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 9]
```diff
@@ -92,6 +97,196 @@ def _init_weights(self, module):
...
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, None
+
+
+def flex_attention_forward(
+    query, key, value, attention_mask, head_mask, norm_factor, output_attentions=False, **_kwargs
+):
+    causal_mask_exists = attention_mask is not None
+    head_mask_exists = head_mask is not None
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Changed it, was just a personal preference

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Yea uh idk what I was thinking; it's ugly af xD changed it up!

File: src/transformers/modeling_flash_attention_utils.py [hunk 10]
```diff
@@ -248,6 +300,11 @@ def _flash_attention_forward(
     if softcap is not None:
         flash_kwargs["softcap"] = softcap
 
+    # PEFT possibly silently casts tensors to fp32, this potentially reconverts to correct dtype or is a no op
+    query_states, key_states, value_states = fa_peft_integration_check(
+        query_states, key_states, value_states, target_dtype
+    )
+
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py:307]: New peft check within the FA2 interface.

File: src/transformers/modeling_utils.py [hunk 11]
```diff
@@ -1799,6 +1809,35 @@ def _check_and_enable_sdpa(cls, config, hard_check_only: bool = False) -> Pretra
...
+        If all checks pass and `hard_check_only` is False, the method will set the config attribute `_attn_implementation` to "flex_attention" so that the model can initialize the correct attention module.
+        """
+        if hard_check_only:
+            if not cls._supports_flex_attn:
+                raise ValueError(
+                    f"{cls.__name__} does not support an attention implementation through torch's flex_attention."
+                    " Please request the support for this architecture: https://github.com/huggingface/transformers/issues/34809."
+                    " If you believe this error is a bug, please open an issue in Transformers GitHub repository"
+                    ' and load your model with the argument `attn_implementation="eager"` meanwhile.'
+                    ' Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="eager")`'
```
Comment (vasqu) [inline src/transformers/modeling_utils.py:1826]: Added the corresponding issue now

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 12]
```diff
@@ -92,6 +95,171 @@ def _init_weights(self, module):
...
+
+    # GPT-neo-X casts query and key in fp32 to apply rotary embedding in full precision
+    value_dtype = value.dtype
+    query = query.to(value_dtype)
+    key = key.to(value_dtype)
+
+    # Avoid torch==2.1.2 specific bug for the memory-efficient backend in SDPA
+    query = query.contiguous()
+    key = key.contiguous()
+    value = value.contiguous()
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:201]: I think it's fine to call contiguous either way! But could add an additional check if wanted

**Review by vasqu — COMMENTED** [review]
I think the PR gets slowly wrapped up; some things that should be made into separate PR(s) imo:
- [ ] Common tests
- [ ] Docs
- [ ] Tracking when attn implementation is manually changed within the config

**Comment by vasqu** [comment]
@ArthurZucker updated per review, looking forward to the next round ;)

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 13]
```diff
@@ -92,6 +95,171 @@ def _init_weights(self, module):
...
+        return score
+
+    attn_output = flex_attention(
+        query,
+        key,
+        value,
+        score_mod=causal_mod,
+        enable_gqa=True,
+        scale=norm_factor,
+        return_lse=output_attentions,
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:243]: I think it we could drop the output attentions here, always return both and let the remaining forward (we call from handle it).

Looked into the torch code, and they also always return both but make an if/else to return just one or both so there shouldn't be any downside imo.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 14]
```diff
@@ -147,20 +315,72 @@ def forward(
...
+            "sdpa",
+            "flash_attention_2",
+        ]:
+            logger.warning_once(
+                f"Setting `attention_type` to `eager` because `{attention_type}` does not support"
+                f" `output_attentions=True` or `head_mask`."
+            )
+            attention_type = "eager"
+
+        if (
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:355]:
```suggestion

        elif (
```

**Review by ArthurZucker — APPROVED** [review]
Very nice! 
Left a few small comments but almost ready to go!

File: src/transformers/modeling_flash_attention_utils.py (hunk 10 above)
Comment (vasqu) [reply src/transformers/modeling_flash_attention_utils.py:300]: I'd leave that to a separate PR, this PR is big enough already :p

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 13 above)
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Added a comment!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 15]
```diff
@@ -147,20 +315,72 @@ def forward(
...
+        if self.config._attn_implementation == "flash_attention_2":
+            input_dtype = value.dtype
+            if input_dtype == torch.float32:
+                if torch.is_autocast_enabled():
+                    target_dtype = torch.get_autocast_gpu_dtype()
+                # Handle the case where the model is quantized
+                elif hasattr(self.config, "_pre_quantization_dtype"):
+                    target_dtype = self.config._pre_quantization_dtype
+                else:
+                    target_dtype = self.query_key_value.weight.dtype
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: The problem is that we are dependent on the models weights in the last case, i.e. `self.query_key_value.weight.dtype` - I moved it to a separate function but it definitely should be deprecated at some point.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 16]
```diff
@@ -92,6 +95,169 @@ def _init_weights(self, module):
...
+    attention_dropout,
+    training,
+    target_dtype=None,
+    **_kwargs,
+):
+    query_length = query.shape[-2]
+
+    # GPT-neo-X casts query and key in fp32 to apply rotary embedding in full precision
+    query = query.to(value.dtype)
+    key = key.to(value.dtype)
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:158]: Tbh, not sure if this is even needed atp. It seems like RoPE implementations changed to a point where it's no longer necessary. Just a heads up, would leave this to another PR as well.

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 17]
```diff
@@ -92,6 +95,169 @@ def _init_weights(self, module):
...
+def sdpa_attention_forward(query, key, value, attention_mask, attention_dropout, training, **_kwargs):
+    q_len = query.shape[-2]
+
+    causal_mask = attention_mask
+    if attention_mask is not None:
+        causal_mask = causal_mask[:, :, :, : key.shape[-2]]
+
+    # GPT-neo-X casts query and key in fp32 to apply rotary embedding in full precision
+    query = query.to(value.dtype)
+    key = key.to(value.dtype)
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:194]: Same here

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 18]
```diff
@@ -92,6 +95,169 @@ def _init_weights(self, module):
...
+        score_mod=causal_mod,
+        enable_gqa=True,
+        scale=norm_factor,
+        # Last time checked on PyTorch == 2.5.1: Flex Attention always computes the lse regardless.
+        # For simplification, we thus always return it as no additional computations are introduced.
+        return_lse=True,
+    )
+
+    # lse is returned in float32
+    attn_weights = attn_weights.to(value.dtype)
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:245]: Reconvert to correct dtype

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 19]
```diff
@@ -147,20 +313,57 @@ def forward(
...
+            query,
+            key,
+            value,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+            norm_factor=self.norm_factor,
+            attention_dropout=self.config.attention_dropout,
+            training=self.training,
+            # Flash Attention 2 specific PEFT check
+            target_dtype=self._fa_peft_dtype_check(value),
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:361]: Like I said in another comment dependent on weights so it's hard to move it to the fa forward without passing additional info (like the weights dtype).

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 20]
```diff
@@ -250,262 +453,47 @@ def _attn_projections_and_rope(
...
+            input_dtype = value.dtype
+            if input_dtype == torch.float32:
+                if torch.is_autocast_enabled():
+                    target_dtype = torch.get_autocast_gpu_dtype()
+                # Handle the case where the model is quantized
+                elif hasattr(self.config, "_pre_quantization_dtype"):
+                    target_dtype = self.config._pre_quantization_dtype
+                else:
+                    target_dtype = self.query_key_value.weight.dtype
+        return target_dtype
```
Comment (vasqu) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:473]: The new function I mentioned.

**Review by vasqu — COMMENTED** [review]
Just a list of TODOs (separate PRs):
- [ ] Deprecate FA peft integration checks
- [ ] Deprecate RoPE reconversion
- [ ] Docs
- [ ] Common Tests
- [ ] Track when attn implementation is manually changed

Otherwise, I think this PR is good now!

**Comment by vasqu** [comment]
@ArthurZucker Hopefully the last round :crossed_fingers:

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py]: Fyi @ArthurZucker https://github.com/pytorch/pytorch/issues/141784

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 18 above)
Comment (ArthurZucker) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:245]: sounds good!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 20 above)
Comment (ArthurZucker) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:473]: thanks

**Review by ArthurZucker — APPROVED** [review]
LGTM now thanks a lot for the refactor!
